src/Fonts.cpp

```diff
@@ -0,0 +1,505 @@
+
+    file.close();
+
+    return true;
+}
+
+void GlyphRangesBuilder::BuildRanges(ImVector<ImWchar>* apRange)
+{
+    m_builder.BuildRanges(apRange);
+    m_needsRebuild = false;
+}
+
+Font::Font(const std::filesystem::path& acPath)
+    : m_path(acPath)
+{
+    m_name = UTF16ToUTF8(m_path.stem().native());
+}
+Font::Font(const std::string& acName, const std::filesystem::path& acPath)
+    : m_name(acName)
+    , m_path(acPath)
+{
+}
+std::string Font::GetName() const
+{
+    return m_name;
+}
+std::filesystem::path Font::GetPath() const
+{
+    return m_path;
+}
+bool Font::Exists()
+{
+    return exists(m_path);
+}
+
+// Build Fonts
+// if custom font not set:
+//     we merge and use all the notosans fonts
+//     (e.g. NotoSans-Regular.ttf + NotoSansJP-Regular.otf + ...).
+//
+// if custom font is set:
+//     we use the custom font (e.g. c:/windows/fonts/Comic.ttf).
+// TODO: Refactor this
+void Fonts::BuildFonts(const SIZE& acOutSize)
+{
+    // TODO: scale also by DPI
+    const auto [resx, resy] = acOutSize;
+    const auto scaleFromReference = std::min(static_cast<float>(resx) / 1920.0f, static_cast<float>(resy) / 1080.0f);
+
+    auto& io = ImGui::GetIO();
+    io.Fonts->Clear();
+
+    const auto& fontSettings = m_options.Font;
+    const float fontSize = std::floorf(fontSettings.BaseSize * scaleFromReference);
+    // scale emoji's fontsize by 0.8 to make the glyphs roughly the same size as the main font. not sure about other font, don't really have a good solution.
+    const float emojiSize = fontSize * 0.8f;
+
+    // get fontpaths
+    // Get custom font paths from options
+    const auto customMainFontPath = GetFontPathFromOption(fontSettings.MainFont);
+    if (customMainFontPath.empty() && !fontSettings.MainFont.empty() && fontSettings.MainFont != "Default")
+        Log::Error("Can't find custom main font at path: {}", GetAbsolutePath(fontSettings.MainFont, m_paths.Fonts(), true).string());
+
+    const auto customMonosFontPath = GetFontPathFromOption(fontSettings.MonoFont);
+    if (customMonosFontPath.empty() && !fontSettings.MonoFont.empty() && fontSettings.MonoFont != "Default")
+        Log::Error("Can't find custom main font at path: {}", GetAbsolutePath(fontSettings.MonoFont, m_paths.Fonts(), true).string());
+
+    const bool useCustomMainFont = !customMainFontPath.empty();
+    const bool useCustomMonosFont = !customMonosFontPath.empty();
+
+    const auto defaultMainFontPath = GetAbsolutePath(m_defaultMainFont, m_paths.Fonts(), false);
+    const auto defaultMonoFontPath = GetAbsolutePath(m_defaultMonoFont, m_paths.Fonts(), false);
+
+    // Set main font path to default if customMainFontPath is empty or doesnt exist.
+    const auto mainFontPath = useCustomMainFont ? customMainFontPath : defaultMainFontPath;
+    if (mainFontPath.empty())
+        Log::Error("Can't find default main font at path: {}, will use built-in font.", GetAbsolutePath(m_defaultMainFont, m_paths.Fonts(), true).string());
```

Comment:
Probably best to trigger warnings if it's not breaking anything and has a fallback

Reply:
Done